multiplex.go

```diff
@@ -104,41 +110,45 @@ func NewMultiplex(con net.Conn, initiator bool, memoryManager MemoryManager) (*M
 	mp := &Multiplex{
 		con:           con,
 		initiator:     initiator,
-		buf:           bufio.NewReader(con),
 		channels:      make(map[streamID]*Stream),
 		closed:        make(chan struct{}),
 		shutdown:      make(chan struct{}),
-		writeCh:       make(chan []byte, 16),
 		nstreams:      make(chan *Stream, 16),
 		memoryManager: memoryManager,
 	}
 
-	// up-front reserve memory for max buffers
-	bufs := 0
-	var err error
-	for i := 0; i < MaxBuffers; i++ {
+	// up-front reserve memory for the essential buffers (1 input, 1 output + the reader buffer)
+	if err := mp.memoryManager.ReserveMemory(3*BufferSize, 255); err != nil {
+		return nil, err
+	}
+
+	mp.reservedMemory += 3 * BufferSize
```

Comment:
Can we specify the `3*BufferSize` in a constant so we don't duplicate it?

Reply:
sure, why not.

Reply:
done.